pkg/periodic/periodic.go

```diff
@@ -0,0 +1,104 @@
+package periodic
+
+import (
+	"context"
+	"sync/atomic"
+	"time"
+)
+
+// Option configures Start.
+type Option interface {
+	apply(*periodic)
+}
+
+// Stoper implements the Stop method,
+// which stops a periodic task from Start().
+type Stoper interface {
+	Stop() // Stops a periodic task.
+}
+
+// Tick is the value for periodic task callbacks that
+// contains the time of the tick and
+// the time elapsed since the start of the periodic task.
+type Tick struct {
+	Elapsed time.Duration
+	Time    time.Time
+}
+
+// OnStop configures a callback that is executed when a periodic task is stopped or canceled.
+func OnStop(f func(Tick)) Option {
+	return optionFunc(func(p *periodic) {
+		p.onStop = f
+	})
+}
+
+// Start starts a periodic task with a ticker at the specified interval,
+// which executes the given callback after each tick.
+// Call Stop() on the return value in order to stop the ticker and to release associated resources.
+// The interval must be greater than zero.
+func Start(ctx context.Context, interval time.Duration, callback func(Tick), options ...Option) Stoper {
+	t := &periodic{
+		interval: interval,
+		callback: callback,
+	}
+
+	for _, option := range options {
+		option.apply(t)
+	}
+
+	ctx, cancelCtx := context.WithCancel(ctx)
+	ticker := time.NewTicker(t.interval)
+	start := time.Now()
+
+	go func() {
+		defer ticker.Stop()
+
+		for {
+			select {
+			case _time := <-ticker.C:
+				t.callback(Tick{
+					Elapsed: _time.Sub(start),
+					Time:    _time,
+				})
+			case <-ctx.Done():
+				if t.onStop != nil {
+					now := time.Now()
+					t.onStop(Tick{
+						Elapsed: now.Sub(start),
+						Time:    now,
+					})
+				}
+
+				return
+			}
+		}
+	}()
+
+	return stoperFunc(func() {
+		if !atomic.CompareAndSwapUint32(&t.stopped, 0, 1) {
```

Comment:
Because to me this looks like this just reimplements `sync.Once`. Or is there a relevant difference in functionality? ("relevant difference" because `sync.Once` also ensures that the function finished executing before any call returns, but as `cancelCtx()` is an asynchronous operation anyways, this doesn't matter much here.)